cpp/src/quantiles/tdigest/tdigest.cu

```diff
@@ -0,0 +1,367 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/tdigest/tdigest.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/structs/structs_column_view.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+// https://developer.nvidia.com/blog/lerp-faster-cuda/
+template <typename T>
+__device__ inline T lerp(T v0, T v1, T t)
+{
+  return fma(t, v1, fma(-t, v0, v0));
+}
+
+template <typename T>
+struct cast_output {
+  data_type t;
+
+  __device__ device_storage_type_t<T> operator()(double v)
+  {
+    if constexpr (!cudf::is_fixed_point<T>()) { return static_cast<T>(v); }
+    if constexpr (cudf::is_fixed_point<T>()) {
```

Comment:
I'm not sure what etiquette demands here, for `if constexpr`. @robertmaynard, would a simple `else` suffice, from the standpoint of unreachable code? Skip even the `else`?
```c++
__device__ device_storage_type_t<T> operator()(double v)
{
    if constexpr (!cudf::is_fixed_point<T>()) { return static_cast<T>(v); }
    return T{v, numeric::scale_type{t.scale()}}.value();
}
```

Reply:
That doesn't work as the compiler still attempts to compile this statement with various non-fixed-point types that don't work with it.

`T{v, numeric::scale_type{t.scale()}}.value();`

Reply:
:)) I knew there was a reason it's phrased this way.